Add new Parent block selector to child blocks (mahjong)

shaunandrews opened the pull request:
This PR aims to create a UI that allows you to select the parent block from the currently selected block's toolbar:

![Toolbar Chip Hover with Delay](https://user-images.githubusercontent.com/191598/77200050-3691a080-6ac0-11ea-8d2b-ec657408c76f.gif)

shaunandrews committed:
First pass.

shaunandrews commented:
Got a quick first pass up that looks like this:

![Parent Selector Draft 1](https://user-images.githubusercontent.com/191598/77200197-78bae200-6ac0-11ea-900e-510bf8901b23.gif)

paaljoachim commented:
Hey Shaun

I really like example 2. With the selecting of the "Column" parent.
It right away let's the user know that it is associated with the block, and will then notice that the parent block is selected.

gziolo commented:
This is cool. Do you have some ideas how user would be able to use the parent button using only keyboard navigation?

ZebulanStanphill commented:
@gziolo I imagine that you would get to it by shift-tabbing from the movers. In other words, you would shift-tab from the block icon to the movers, and then again to get to the select-parent button. That would make the most sense from a visual-order-should-match-DOM-order perspective.

Right now, there is an inserter button when you shift-tab from the movers, but I think that's a poor place to put that control. See #13571.

shaunandrews committed:
Adding the BlockIcon

shaunandrews commented:
Here's the latest on this PR:

![Parent Block Selector Demo i2](https://user-images.githubusercontent.com/191598/77257740-9ae16b00-6c4c-11ea-8ac8-5ebf60112696.gif)

Things feel pretty good, though I could use some 👀 on the code — I'm certain there are missing things or better ways to accomplish some of these things.

> Do you have some ideas how user would be able to use the parent button using only keyboard navigation?

You can put `:focus` on the Parent button using the normal tabbing — the button appears between the movers and the current block icon; When focused, the Parent button appears.

It's also worth noting that you can already use the keyboard's up/down arrows to move through the block hierarchy.

ZebulanStanphill commented:
I think that it would make more sense if the tab order was select-parent -> movers -> block switcher rather than movers -> select-parent -> block switcher. The select-parent button could be viewed as a sort of navigation link somewhat outside of the rest of the block toolbar (hence why it is floating a bit away from the rest of the toolbar). It's also the first thing in the visual order if you go top-bottom, left-right, so from an accessibility perspective, it should be the first thing in the DOM order.

gziolo commented:
> I think that it would make more sense if the tab order was select-parent -> movers -> block switcher rather than movers -> select-parent -> block switcher. The select-parent button could be viewed as a sort of navigation link somewhat outside of the rest of the block toolbar (hence why it is floating a bit away from the rest of the toolbar). It's also the first thing in the visual order if you go top-bottom, left-right, so from an accessibility perspective, it should be the first thing in the DOM order.

I agree that it makes sense to have the select parent button first. It's also a regular button so it can't be in the tab order in the middle of the toolbar. Keep in mind that we still didn't finish refactoring all tollbars to handle toolbar buttons navigation with arrow keys only and making the toolbar being a single tab stop. I guess for the accessibility reasons the parent selector should be always focusable even when it is not visible? I don't think it's implemented at the moment.

youknowriad commented:
Should we support multi-selection too? I believe it triggers an error now.

jasmussen commented:
Rocking it, Shaun! This is working better than I expected:

![vertical](https://user-images.githubusercontent.com/1204802/77298267-f8c49000-6cea-11ea-95a3-09958e99e8d1.gif)

Two things:

1. The space between the "go up" button and the toolbar should be 12px, i.e. $grid-unit-15.
2. I would make the top button appear in lockstep with the mover control, not shortly after.

> Should we support multi-selection too? I believe it triggers an error now.

I would disable it entirely for multi selection.

Tab order feels correct to me, but I will defer to those with stronger opinions and background on best practices.

mtias commented on packages/block-editor/src/components/block-parent/index.js:
```diff
@@ -0,0 +1,71 @@
+          <Button
+            className="block-editor-block-parent__button"
+            onClick={ () => selectBlock( firstParentClientID ) }
+            label={ __( 'Select parent block' ) }
```
I'd try `Select {parentBlockName}`, as the icon alone is not sufficient information.

shaunandrews replied:
Yup yup, I had to figure out how to put a variable in the translate function. Found some random examples and got it working.

mtias commented on packages/block-editor/src/components/block-parent/index.js:
```diff
@@ -0,0 +1,71 @@
+ * Internal dependencies
+ */
+import BlockIcon from '../block-icon';
+import BlockTitle from '../block-title';
```
This doesn't seem to be used in the file.

ZebulanStanphill commented:
>I would make the top button appear in lockstep with the mover control, not shortly after.

Agreed.

shaunandrews committed:
Adjust the position to match the overall grid.

shaunandrews committed:
Removing the delay on the Parent button transition.

shaunandrews committed:
Update the tooltip to display the parent block's title.

Cleaning a bunch of stuff from pre-commit hooks I had been ignoring.

shaunandrews committed:
Adjust code indentation.

shaunandrews committed:
Don't show the Parent button if multiple blocks are selected.

shaunandrews committed:
Some auto spacing nonsense.

shaunandrews commented:
![Parent Block Selector Demo i3](https://user-images.githubusercontent.com/191598/77449735-b6718080-6dc8-11ea-9c4c-4278ff6573b9.gif)

Pushed a handful of changes:

- Updated the space between the Parent button and the Toolbar to 12px.
- The Parent button now appears at the same time as the Movers.
- The tooltip now displays the parent block's title.
- There was a bunch of prettier stuff that I keep ignoring (`--no-verify` is a nasty habit); I've cleaned things up now in the code.
- The Parent button doesn't render if multiple blocks are selected. This could use some more testing, but I no longer see an error when selecting multiple blocks.

ZebulanStanphill commented on packages/block-editor/src/components/block-parent/index.js:
```diff
@@ -0,0 +1,69 @@
+					className="block-editor-block-parent__button"
+					onClick={ () => selectBlock( firstParentClientID ) }
+					label={ sprintf(
+						__( 'Select %s' ),
```
I think we should make it more clear that you're selecting the parent:
```suggestion
						__( 'Select parent: %s' ),
```

ZebulanStanphill commented:
Just tested out this PR. This is fantastic! Already, I can see myself using this button far more often than the footer breadcrumb links. The only thing I'd change is the label (see code suggestion I recently made) and the DOM/tab order of the button. Aside from that, I think this is perfect.

gziolo commented:
It works great, it's going to be awesome 💯 

I second all the concerns @ZebulanStanphill shared. I recorder a screencast where I use the mix of tabbing and arrow keys (left/right) to navigate to parent blocks:

![block-parent](https://user-images.githubusercontent.com/699132/77623244-0d4d9600-6f40-11ea-8166-20b1a083dc12.gif)

One thing I'm not sure about is whether this new "select parent" button should be part of the block toolbar and thus reachable with arrow left/right keys. It feels strange when you see how the focus moves at the moment because it's disconnected from the rest of the toolbar by both the spacing used and the direction that goes to the top. Maybe @enriquesanchez or @MarcoZehe could advise here on the best approach here.

gziolo commented on packages/block-editor/src/components/block-parent/index.js:
```diff
@@ -0,0 +1,69 @@
+
+	const firstParentClientID = parents[ parents.length - 1 ];
+
+	const { parentBlockType } = useSelect( ( select ) => {
```
You can consolidate this `useSelect` call with the previous one and expose only data used:
- `firstParentClientID`
- `parentBlockTypeTitle`
- `parentBlockTypeIcon`

gziolo commented on packages/block-editor/src/components/block-parent/index.js:
```diff
@@ -0,0 +1,69 @@
+		};
+	}, [] );
+
+	if ( parents && parents.length ) {
```
Would it work as follows:
```suggestion
	if ( firstParentClientID ) {
```

It should simplify the code.

gziolo commented on packages/block-editor/src/components/block-toolbar/index.js:
```diff
@@ -85,6 +86,12 @@ export default function BlockToolbar( { hideDragHandle } ) {
+	const animatedBlockParentStyles = {
+		opacity: shouldShowMovers ? 1 : 0,
+		transform: shouldShowMovers ? 'translateY(-63px)' : 'translateY(0)',
+		transitionDelay: shouldShowMovers ? 0 : 0,
```
It can be simplified to:
```suggestion
		transitionDelay: 0,
```

paaljoachim commented:
I took a test run using gutenberg.run.

![Screen Shot 2020-03-26 at 09 28 18](https://user-images.githubusercontent.com/5323259/77625739-69b2b480-6f44-11ea-8431-1f5b3512dba4.png)

It would be nicer to bring the popup "parent" icon closer to the toolbar. Here is a mockup. To give a feeling for how the parent icon looks when it is closer to the toolbar. It creates a closeness between objects. 

![Parent-icon-closer-to-toolbar](https://user-images.githubusercontent.com/5323259/77626691-03c72c80-6f46-11ea-845f-456811cf5388.jpg)

ZebulanStanphill commented:
I think the select-parent button should feel a little detached from the toolbar because it isn't part of the toolbar. I think that was the idea in the original mockup. It's sort of like the sibling inserter in that respect. The button could also be viewed as a sort of navigational skip link, similar to the one used to get to the Inspector. That's one of the reasons I think it should be the "leftmost" tabbable. Having it accessible in the middle of the toolbar doesn't make sense if it isn't part of the toolbar (both from a logical and visual perspective).

shaunandrews committed:
Moving the parent button to change its tab order.

shaunandrews committed:
Clean up the spacing in the code to make the code-gods happy.

ZebulanStanphill commented:
Nice! Alright, there are two more issues I've spotted, functionality-wise.

First, the parent-selector appears underneath the top toolbar when the block toolbar is sticky. This might be solvable using z-index tweaks.
![image](https://user-images.githubusercontent.com/19592990/77683874-19991980-6f67-11ea-94df-6677215dcb2f.png)

When the "Top toolbar" option is enabled, the parent-selector is placed off-screen. In this case, I guess the solution would be to simply place the button to the left of the movers, or else omit the button and add an equivalent in the Block Navigation dropdown?
![image](https://user-images.githubusercontent.com/19592990/77684003-4c431200-6f67-11ea-90d3-6c7522e36512.png)

enriquesanchez commented:
>The button could also be viewed as a sort of navigational skip link, similar to the one used to get to the Inspector. That's one of the reasons I think it should be the "leftmost" tabbable. Having it accessible in the middle of the toolbar doesn't make sense if it isn't part of the toolbar (both technically and visually.

I just tested and I agree with @ZebulanStanphill. If we look at keyboard navigation as a linear top-to-bottom flow, the parent skip link should be the top/left most tabbale item.

ellatrix commented:
> First, the parent-selector appears underneath the top toolbar when the block toolbar is sticky.

I wanted to raise this concern as well. How can it be solved?

jasmussen commented:
> First, the parent-selector appears underneath the top toolbar when the block toolbar is sticky.

This is a good challenge, and it's a challenge we also have with full site editing where some blocks will put themselves tightly against the top of the screen. 

In the past we discussed adding clearance all around the canvas to allow for this toolbar. But the point of the sticky toolbar and the parent selector shows that this solution would not work for that.

One solution we can do is to accept the button overlapping the top toolbar. In the following I used the web-inspector to verify whether it was possible for the block toolbar to overlap the top toolbar. It is:

<img width="541" alt="Screenshot 2020-03-27 at 10 27 42" src="https://user-images.githubusercontent.com/1204802/77741831-d34ec380-7015-11ea-8ce3-d6d41e3cb4ed.png">

For that to work, `.edit-post-visual-editor > .popover-slot` needs to receive both `position: absolute;`, and a `z-index: 30;` or higher.

mtias commented:
I'd be inclined to not render this for the top toolbar setting since you'd be already close to the navigator in the header when that's your configuration. We still need to handle floating toolbar when it's close to the top, though.

shaunandrews committed:
Cleaning up some CSS.

shaunandrews committed:
Using a className for shouldShowMovers instead of the inline styles.

shaunandrews committed:
Hide the Parent button in Top Toolbar mode.

shaunandrews committed:
Moving the z-index for the block toolbar so it renders above the top bar. Adjusting the position so it appears centered in the top bar.

shaunandrews commented:
I've hidden the Parent button when Top Toolbar mode is active.

Also, I've updated the `z-index` of the floating toolbar to be above the top bar. I thought I would hate this, but it doesn't bother me at all in practice. Here's how it looks:

![Parent Button z-index](https://user-images.githubusercontent.com/191598/78294824-03f09a80-74f9-11ea-969f-6fc0572509cf.gif)

jasmussen commented:
I'm excited about this, Shaun! 

Can you _try_ to make the space between the parent button and the toolbar be 12px? ($grid-unit-15) — in my minds eye, while it looks like a big bit of space, it will align it perfectly to the grid, and given it animates out, the connection feels there. 

I always try and design from a principle of: if you're going to so something, _do it_. Which means, either a space between the two components that is _intentional_, or do _no space at all. (I'm not a fan of the latter as it will create a cross-shape) but hopefully it conveys my thought process.

mtias commented:
Agreed with the intentional spacing. Looking forward to this one!

jasmussen commented:
> mtias changed the title Add new Parent block selector to child blocks Add new Parent block selector to child blocks (mahjong) 

(•_•)
  ( •_•)>⌐■-■
     (⌐■_■)

ZebulanStanphill commented on packages/block-editor/src/components/block-toolbar/style.scss:
```diff
@@ -183,6 +183,25 @@
+
+	.is-showing-movers & {
+		opacity: 1;
+		transform: translateY(-54px);
```
I think this change will apply the spacing @jasmussen wants, while also avoiding hardcoded values:
```suggestion
		transform: translateY(-($block-toolbar-height + $grid-unit-15));
```

shaunandrews replied:
Ah, thanks. I added this.

shaunandrews commented:
I had the 12px spacing, and reduced it _intentionally_ so that the the button appears vertically centered when it overlaps the top bar. Its probably not the best scenario to design for, but with the 12px, the Parent button looks a little strange:

<img width="741" alt="image" src="https://user-images.githubusercontent.com/191598/78368067-0dc5dc80-7591-11ea-9ee2-9d160fc029f0.png">

I was hoping I could adjust the spacing _only_ when the toolbar is "stuck" to the top, but there's not className for me to use.

jasmussen commented:
I think that's an acceptable edgecase to start with and then revisit if it becomes common enough to annoy you!

ZebulanStanphill commented:
@jasmussen So are you suggesting we go with the standard spacing or not? I can't really tell from your comment.

shaunandrews committed:
Use the CSS variables for the spacing.

shaunandrews committed:
Update the tooltip to use the word "parent" at a hope of making it more helpful.

shaunandrews commented:
I updated the spacing to use the CSS variable and added the word "parent" to the tooltip:

<img width="420" alt="image" src="https://user-images.githubusercontent.com/191598/78588770-30fcc000-780d-11ea-9df4-d9b23f411ab2.png">

ZebulanStanphill commented:
I tested this PR via http://gutenberg.run and there seems to be a bit of a styling issue at the moment:
![image](https://user-images.githubusercontent.com/19592990/78596370-483dac80-7811-11ea-8739-42d78b09290d.png)

johnstonphilip commented:
I hate to throw another idea at you at this point in a PR, but what if the Block Navigator popped up instead?

<img width="593" alt="Screen Shot 2020-04-08 at 8 52 33 AM" src="https://user-images.githubusercontent.com/7538525/78786592-03338a80-7977-11ea-93a2-8b4f161b5a6a.png">

**Pros:**
- You can navigate anywhere in your document easily, instead of just a single container up
- Users don't have to learn another interface for block navigation (this PR is the 3rd after breadcrumbs and Block navigation)
- It's easier to understand what is going on (in my opinion). With this PR, the single icon popping up doesn't really tell me very much. Whereas the Block navigation tells me a lot more.

**Cons:**
- It takes up more space.

ZebulanStanphill commented:
@johnstonphilip Unfortunately, I don't think that idea scales very well. Keep in mind, this appears on hover, so every time you would hover over the block icon (often just to move the block using the movers or use the block type/style switcher), you would get this huge thing suddenly eating up a large portion of the screen.

johnstonphilip commented:
@ZebulanStanphill I hear that, and it makes sense. I'm not totally convinced this PR is the right move either. Nested blocks are truly one of the hardest things to solve. I know someone mentioned Layers from Divi, but that concept is something I was never able to wrap my brain around either.

shaunandrews commented:
@ZebulanStanphill I'm unable to reproduce this; Could you tell me if you still see it, and maybe your browser/OS info?

ZebulanStanphill commented:
@shaunandrews I'm using a Chromium 80-based browser on Arch Linux. I'm testing the branch via [gutenberg.run](http://gutenberg.run/). Have you tried clearing the cache in your browser? I cleared mine but the issue persisted.

ZebulanStanphill commented:
Instead of calling this component `BlockParent`, I think it should be called `BlockParentSelector`. (And the class names should be updated accordingly.) It's not the block parent itself, just a tool to select it.

tomjn commented:
It seems a bit random having an additional icon above the transform button, could we add a visual affordance indicating it's for the parent? Like an arrow going to the left then up? Similar to how the transform button has the little recycle symbol? The selection tool cursor might be a good indicator

Right now it's not possible to tell what the button does just by looking at it, interaction of some kind is necessary, be that hovering or clicking

shaunandrews committed:
Renaming the component to be more descriptive by adding "Selector".

This component isn't the block's parent, its a way to select the block's parent.

ZebulanStanphill commented:
@tomjn Just to be clear, the "recycle" icon thing was removed from the block type/style switcher during the G2 UI changes.

I agree some kind of indicator of what the button does might be useful, but I'm not sure what to use. An upwards arrow/chevron seems like the most obvious choice. We should try to avoid making the button look too cluttered in the process, though.

shaunandrews commented:
> I'm testing the branch via gutenberg.run.

From what I can tell this only happens on Gutenberg.run — it seems like it might be running a different/older version of WordPress Core. I'm looking to nail down the styles that are causing the problem, but haven't found it yet.

> I think it should be called BlockParentSelector.

Agreed. Renamed.

> An upwards arrow/chevron seems like the most obvious choice. We should try to avoid making the button look too cluttered in the process, though.

The position of the button was intended to be an indicator, but I've tried adding an arrow and an line to help reinforce. I'm not sure either really helps.

<img width="410" alt="image" src="https://user-images.githubusercontent.com/191598/79488545-e129b000-7fe7-11ea-8fa4-904e7bd8afb0.png">

For now, I'd love to see how people use this before adding more to the UI.

ZebulanStanphill commented:
I think the style issue is due to the lack of a wrapping `.components-toolbar`. There's a style rule for `.components-toolbar .components-button.has-icon` that is not being applied to the parent selector.

And yeah, I'd be happy if this merged as-is (assuming the size issue gets fixed). We can iterate on the design if necessary in follow-up PRs.

mtias commented:
> For now, I'd love to see how people use this before adding more to the UI.

Agreed

shaunandrews committed:
Updating the padding a height of the parent button so it matches the size of other toolbar buttons.

shaunandrews commented:
Fixed the issue @ZebulanStanphill pointed out — the parent button should now be the right size.

ZebulanStanphill reviewed (approved):
Due to the increased z-index on the block toolbar, this can happen now:
![image](https://user-images.githubusercontent.com/19592990/79793003-7e048a00-8315-11ea-938e-53214a399457.png)

Are we okay with this behavior?

Also, I'm not sure if the focus outline on the parent-selector should match the block toolbar or not. I think it's fine as-is since the "inside" outlines seem to only be used for toolbar buttons, and this button isn't actually part of the block toolbar.

![image](https://user-images.githubusercontent.com/19592990/79793763-ba84b580-8316-11ea-9aea-0344ba6d3b47.png)

Approving because I think, aside from the minor issues mentioned above, this is ready to merge.

shaunandrews committed:
Making Travis CI happy.

youknowriad commented on packages/block-editor/src/components/block-parent-selector/index.js:
```diff
@@ -0,0 +1,70 @@
+		return {
+			parentBlockType: getBlockType( parentBlockName ),
+		};
+	}, [] );
```
Why do we have two different useSelect calls? can we just use a single one?

youknowriad commented on packages/block-editor/src/components/block-parent-selector/style.scss:
```diff
@@ -0,0 +1,10 @@
+	border: 1px solid $dark-gray-primary;
+	border-radius: $radius-block-ui;
+
+	.block-editor-block-parent-selector__button {
```
Is the nesting needed here?

shaunandrews replied:
Yes, the specificity allows use to override the margin on the button. Without that, the button is misaligned.

youknowriad commented on packages/block-editor/src/components/block-switcher/index.js:
```diff
@@ -138,6 +138,7 @@ export class BlockSwitcher extends Component {
 								label={ label }
 								onKeyDown={ openOnArrowDown }
 								showTooltip
+								tooltipPosition="bottom"
```
Was this intended? I'm not sure it works well in all situations (mobile, top toolbar, full-width blocks...)

shaunandrews replied:
Ah, I forgot about this... I added it because at one point the tooltip wasn't showing without it. Now it seems that the tooltip doesn't show regardless. There some conflict being caused by the `<Button />` component in the `block-parent-selector` and I'm unsure how to go about fixing this...

youknowriad commented on packages/block-editor/src/components/block-toolbar/style.scss:
```diff
@@ -64,6 +64,25 @@
+	top: -1px;
+	left: -1px;
+	opacity: 0;
+	transition: all 60ms linear;
```
Should we add the reduced motion support?

shaunandrews replied:
Ah yea. Added it.

youknowriad reviewed (commented):
This is looking good overall, left some small notes.

paaljoachim commented:
Even though this PR when merged will help. 
We also need a hit zone for selecting nested blocks as well as columns.
As the older version of Gutenberg plugin there was a more obvious hit zone to the left of the blocks. Today it is not so obvious making it more difficult in selecting nested or columns.
https://github.com/WordPress/gutenberg/issues/22371

mtias commented:
@paaljoachim the hit-zone wasn't very good because it implied a visual gap that might not exist and made things not render accurately.

paaljoachim commented:
Hey @mtias 

Bottom line is having a direct manipulation when clicking a block in addition to having the toolbar icons being added through this PR. As it is a lot quicker clicking inside certain areas or other ways to do a direct selection.

At the moment we have a small hit zone for columns. Making it possible to do a direct selection. One needs to click outside the rectangle and inside the blue outline. 
![Screen Shot 2020-05-15 at 15 18 09](https://user-images.githubusercontent.com/5323259/82054492-51972000-96bf-11ea-9271-44d25648939c.png)

ZebulanStanphill commented:
I think that, to make selecting blocks up and down the nested hierarchy easier, we're really going to need to make the Block Navigation interface more powerful (and once we do that, we can remove the breadcrumb bar). See #22113, #22287, and #22297.

Adding padding to nested blocks is something that I used to be in favor of back when it was first suggested and implemented, but after having actually used it, I'm glad it was reverted.

ZebulanStanphill commented on packages/block-editor/src/components/block-parent-selector/index.js:
```diff
@@ -0,0 +1,70 @@
+ * Block parent selector component, displaying the hierarchy of the
+ * current block selection as a single icon to "go up" a level.
+ *
+ * @return {WPElement} Block Breadcrumb.
```
```suggestion
 * @return {WPElement} Block parent selector button.
```

ZebulanStanphill commented on packages/block-editor/src/components/block-parent-selector/index.js:
```diff
@@ -0,0 +1,70 @@
+		const selectedBlockClientId = getSelectedBlockClientId();
+		return {
+			parents: getBlockParents( selectedBlockClientId ),
+			clientId: selectedBlockClientId,
+			hasSelection: !! getSelectionStart().clientId,
```
These are never used.

shaunandrews replied:
Ah, thanks. Removed.

ZebulanStanphill commented on packages/block-editor/src/components/block-parent-selector/index.js:
```diff
@@ -0,0 +1,70 @@
+/**
+ * WordPress dependencies
+ */
+import { Button } from '@wordpress/components';
+import { useSelect, useDispatch } from '@wordpress/data';
+import { __, sprintf } from '@wordpress/i18n';
+import { getBlockType } from '@wordpress/blocks';
```
Nitpick: I prefer when imports are in alphabetical order.
```suggestion
import { getBlockType } from '@wordpress/blocks';
import { Button } from '@wordpress/components';
import { useDispatch, useSelect } from '@wordpress/data';
import { __, sprintf } from '@wordpress/i18n';
```

ZebulanStanphill commented on packages/block-editor/src/components/block-parent-selector/index.js:
```diff
@@ -0,0 +1,70 @@
+ *
+ * @return {WPElement} Block Breadcrumb.
+ */
+const BlockParentSelector = function() {
```
You could just do `export default function BlockParentSelector() {` here.

paaljoachim commented:
Zeb those issues you mentioned will help. We need to keep the breadcrumbs bar in the bottom though, as it becomes one of a few ways in how we can select blocks. We will see how this goes.

ZebulanStanphill commented on packages/block-editor/src/components/block-parent-selector/index.js:
```diff
@@ -0,0 +1,70 @@
+		};
+	}, [] );
+
+	const firstParentClientID = parents[ parents.length - 1 ];
```
This should be `firstParentClientId` (note the capitalization). "ID" is not the right way to capitalize it; it's not an acronym. There's an existing precedent, e.g. `getSelectedBlockClientId` or `useInstanceId`.

ZebulanStanphill reviewed (changes requested):
[This behavior still feels janky](https://github.com/WordPress/gutenberg/pull/21056#pullrequestreview-396773721), but I don't know what the solution is there. I do have one idea: if the [movers were changed to always appear in the toolbar](https://github.com/WordPress/gutenberg/issues/21935), we could move the parent selector to the position that the movers are currently at. But of course, that idea depends on another idea being implemented. For now, I'm not sure what the right solution is.

Aside from that, it looks like all this PR needs is some cleanup, polish, and a rebase.

shaunandrews commented:
I hope to return to this branch next week. Thanks for all the feedback.

shaunandrews committed:
Cleaning up some unused code, and fixing some copy/paste mistakes in the code comments.

shaunandrews committed:
Moving some padding around so the focus state of the button appears more like other toolbar buttons.

shaunandrews committed:
Making the d lowercase.

ZebulanStanphill commented on packages/block-editor/src/components/block-parent-selector/index.js:
```diff
@@ -0,0 +1,64 @@
+ */
+export default function BlockParentSelector() {
+	const { selectBlock } = useDispatch( 'core/block-editor' );
+	const { parents } = useSelect( ( select ) => {
+		const { getSelectedBlockClientId, getBlockParents } = select(
+			'core/block-editor'
+		);
+		const selectedBlockClientId = getSelectedBlockClientId();
+		return {
+			parents: getBlockParents( selectedBlockClientId ),
+		};
+	}, [] );
```
Since we're only using `useSelect` for a single value, we can just return it and avoid having to destructure:
```suggestion
	const parents = useSelect( ( select ) => {
		const { getSelectedBlockClientId, getBlockParents } = select(
			'core/block-editor'
		);
		const selectedBlockClientId = getSelectedBlockClientId();
		return getBlockParents( selectedBlockClientId );
	}, [] );
```

ZebulanStanphill commented on packages/block-editor/src/components/block-parent-selector/index.js:
```diff
@@ -0,0 +1,64 @@
+ * Block parent selector component, displaying the hierarchy of the
+ * current block selection as a single icon to "go up" a level.
+ *
+ * @return {WPElement} Parent Block Selector.
```
Is this supposed to be title-cased? I would think it should be sentence case.

```suggestion
 * @return {WPElement} Parent block selector.
```

ZebulanStanphill commented on packages/block-editor/src/components/block-parent-selector/index.js:
```diff
@@ -0,0 +1,64 @@
+		};
+	}, [] );
+
+	if ( parents && parents.length ) {
```
```suggestion
	if ( parents?.length ) {
```

ZebulanStanphill commented on packages/block-editor/src/components/block-parent-selector/index.js:
```diff
@@ -0,0 +1,64 @@
+ */
+export default function BlockParentSelector() {
+	const { selectBlock } = useDispatch( 'core/block-editor' );
+	const { parents } = useSelect( ( select ) => {
+		const { getSelectedBlockClientId, getBlockParents } = select(
+			'core/block-editor'
+		);
+		const selectedBlockClientId = getSelectedBlockClientId();
+		return {
+			parents: getBlockParents( selectedBlockClientId ),
+		};
+	}, [] );
+
+	const firstParentClientId = parents[ parents.length - 1 ];
+
+	const { parentBlockType } = useSelect( ( select ) => {
+		const { getBlockName } = select( 'core/block-editor' );
+		const parentBlockName = getBlockName( firstParentClientId );
+		return {
+			parentBlockType: getBlockType( parentBlockName ),
+		};
+	}, [] );
```
As @youknowriad suggested, we can combine the `useSelect` calls.

```suggestion
	const { parents, parentBlockType } = useSelect( ( select ) => {
		const {
			getBlockName,
			getBlockParents,
			getSelectedBlockClientId
		} = select( 'core/block-editor' );
		const selectedBlockClientId = getSelectedBlockClientId();
		const _parents = getBlockParents( selectedBlockClientId )
		const firstParentClientId = _parents[ _parents.length - 1 ];
		const parentBlockName = getBlockName( firstParentClientId );

		return {
			parents: _parents,
			parentBlockType: getBlockType( parentBlockName ),
		};
	}, [] );
```

shaunandrews replied:
I tried this one, but it breaks the editor. There's a `firstParentClientId not defined" error in the console.

ZebulanStanphill replied:
Whoops! I forgot that `firstParentClientId` was being used outside of the select callback.

This should work:
```suggestion
	const {
		parents,
		parentBlockType,
		firstParentClientId
	} = useSelect( ( select ) => {
		const {
			getBlockName,
			getBlockParents,
			getSelectedBlockClientId
		} = select( 'core/block-editor' );
		const selectedBlockClientId = getSelectedBlockClientId();
		const _parents = getBlockParents( selectedBlockClientId )
		const _firstParentClientId = _parents[ _parents.length - 1 ];
		const parentBlockName = getBlockName( _firstParentClientId );

		return {
			parents: _parents,
			parentBlockType: getBlockType( parentBlockName ),
			firstParentClientId: _firstParentClientId
		};
	}, [] );
```

shaunandrews committed:
Sentence case, and a short way to do the check.

youknowriad commented on packages/block-editor/src/components/block-parent-selector/index.js:
```diff
@@ -0,0 +1,64 @@
+ */
+export default function BlockParentSelector() {
+	const { selectBlock } = useDispatch( 'core/block-editor' );
+	const { parents } = useSelect( ( select ) => {
```
Seems  like you only  need a boolean "hasParents". When using useSelect,  it's better to limit the result to things that change less often. I assume "parents" here is a block list, it changes way too often compared to a boolean (that's a performance trick)

ZebulanStanphill replied:
`parents` is being used to calculate `firstParentClientId`, so it is needed. However, if we combine the `useSelect` calls, we don't have to return `parents` directly; we could calculate `hasParents` in the select callback and return it, and thus simplify `if ( parents?.length )` to `if ( hasParents )`.

shaunandrews replied:
I removed `parent` from the `useSelect` return, and instead used the `firstParentClientId` to do the check. Maybe this is a terrible idea?

ZebulanStanphill replied:
Is `firstParentClientId` a string or a number? If it's a number, is it possible for `firstParentClientId` to be something like `0`, which would be coerced to `false`? I think for the sake of clarity and avoiding issues like that, it may be best to just use a `hasParents` boolean.

Alternatively, you could explicitly check if it's `undefined`:
```js
if ( firstParentClientId !== undefined ) {
```

shaunandrews replied:
I added the explicit `undefined` check.

shaunandrews committed:
Reducing down to a single useSelect.

shaunandrews committed:
Using the `firstParentClientId` rather than `parents` to do the return check.

ZebulanStanphill commented on packages/block-editor/src/components/block-parent-selector/index.js:
```diff
@@ -0,0 +1,60 @@
+			getSelectedBlockClientId,
+		} = select( 'core/block-editor' );
+		const selectedBlockClientId = getSelectedBlockClientId();
+		const _parents = getBlockParents( selectedBlockClientId );
+		const _firstParentClientId = _parents[ _parents.length - 1 ];
```
No need to prefix `parents` with an underscore since it's not conflicting with anything in the upper scope:
```suggestion
		const parents = getBlockParents( selectedBlockClientId );
		const _firstParentClientId = parents[ parents.length - 1 ];
```

shaunandrews committed:
No need to prefix parents with an underscore since it's not conflicting with anything in the upper scope.

Co-authored-by: Zebulan Stanphill <zebulanstanphill@protonmail.com>

shaunandrews committed:
Adding an explicit undefined check.

ZebulanStanphill reviewed (approved):
It looks like there are some code comments from @youknowriad that still need to be addressed. But aside from those, everything looks good to me and seems to be working.

The z-index toolbar overlap issue is still a thing, but I'm willing to accept it to get this feature in sooner.

![image](https://user-images.githubusercontent.com/19592990/79793003-7e048a00-8315-11ea-938e-53214a399457.png)

Approving under the assumption the things youknowriad pointed out will be addressed prior to merge.

shaunandrews committed:
Adding the reduce-motion mixin to respect user settings around animations.

youknowriad commented on packages/block-editor/src/components/block-parent-selector/index.js:
```diff
@@ -0,0 +1,60 @@
+		};
+	}, [] );
+
+	if ( firstParentClientId !== undefined ) {
```
Should we  invert the condition and return early, it's a  bit better in terms of readability

shaunandrews committed:
This shouldn't be needed.

shaunandrews commented:
After the last rebase, I noticed that the tooltip for the Block Switcher component doesn't display when the block is within a parent block.

![tooltips](https://user-images.githubusercontent.com/191598/82584956-3fa3fa00-9b63-11ea-9457-67a8ae437426.gif)

However, for blocks without a parent the tooltip appears to display normally:

<img width="445" alt="image" src="https://user-images.githubusercontent.com/191598/82585025-55b1ba80-9b63-11ea-850c-a7dd4ea0e9b7.png">

And for the life of me, I can't figure out why...

ZebulanStanphill commented on packages/block-editor/src/components/block-parent-selector/index.js:
```diff
@@ -0,0 +1,60 @@
+ * Block parent selector component, displaying the hierarchy of the
+ * current block selection as a single icon to "go up" a level.
+ *
+ * @return {WPElement} Parent block selector.
```
I'm pretty sure this should be `WPComponent`, not `WPElement`?

shaunandrews committed:
I've been told this should be a Component not an Element.

ItsJonQ commented:
> After the last rebase, I noticed that the tooltip for the Block Switcher component doesn't display when the block is within a parent block.

@shaunandrews Haii! I just took a looksy.

I'm not too sure why either 🤔 

I took a look at the `BlockSwitcher` and `BlockParentSelector` blocks. Haven't found anything yet.

My **hunch** tells me that there's probably a `event.stopPropagation()` or something happening somewhere, which is preventing the `Tooltip` from firing.

Commit:
Fix BlockSwitch tooltip

ItsJonQ commented:
@shaunandrews I think I got it!

<img width="348" alt="Screen Shot 2020-05-25 at 10 59 20 AM" src="https://user-images.githubusercontent.com/2322354/82824525-82254980-9e77-11ea-99bd-c02e0905f4e4.png">

I pushed up my change. I followed the trail of props. It looked like the inner `ToolbarButton/Button` was looking for `title` to be used for it's Tooltip (not `label`)

Let me know if that works for you! 🙏

ZebulanStanphill commented:
@ItsJonQ Well that seems like a bug (outside of this PR); it seems like `label` _should_ work, shouldn't it?

ItsJonQ commented:
@ZebulanStanphill From what I can tell, it looks like `ToolbarButton` was updated with some other Toolbar components recently.

cc'ing @diegohaz !

Not saying you broke anything 🙏 ! Just curious. It looks like the `label` property was swapped with `title`. Was wondering if you had any context?

From a semantic perspective, `title` makes more sense. Just confirming the behaviour.

Thanks!

youknowriad commented:
if it's a recent change (swapping label with title), then it's a breaking change and should be undone. Also consistency with the "Button" component seems like a good thing isn't it?

youknowriad reviewed (approved):
It seems like there's a JS unit test failure.

Other than that, it seems like we can land this.

diegohaz commented:
As far as I know, the `ToolbarButton` prop has always been `title`, and it's passed to the `label` prop on `Button`.

These props are named differently on `ToolbarButton` and `Button`:
```jsx
<ToolbarButton title="title" isActive isDisabled />
<Button label="label" isPressed disabled />
```

Besides that, to pass additional props like event handlers to the underlying `Button` when using `ToolbarButton`, you have to use an `extraProps` prop:

```jsx
<ToolbarButton extraProps={ { onKeyDown: ... } } />
<Button onKeyDown={ ... } />
```

I'm not aware of the reasons behind that. But it seems to me that we should deprecate those different props on `ToolbarButton` and adopt the same API as `Button` for consistency.

With #18931, `ToolbarButton` started to also accept the `Button` API, but only when using it as a child of `<Toolbar __experimentalAccessibilityLabel />`.

Commit:
Update BlockSwitch test snapshot

ItsJonQ commented:
> It seems like there's a JS unit test failure.

@youknowriad Wonderful! I just pushed an update to the test snapshots. That should do it

ItsJonQ commented:
Thanks for that context @diegohaz !!

ItsJonQ commented:
Looks like Travis is happy! I'll merge it up.
Thanks for the feedback + review all!

mapk commented:
Really great to see this getting merged! I'm excited to see it in play. 🎉

mtias commented:
Kudos @shaunandrews !

shaunandrews commented:
What a nice way to come back from a week offline — thanks for the assist, @ItsJonQ!
